keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -49,120 +37,52 @@ class EnvironmentPromotionChecker(
       .forEach { (artifact, versions) ->
         if (versions.isEmpty()) {
           log.warn("No versions for ${artifact.type} artifact ${artifact.name} are known")
-        } else {
-          deliveryConfig.environments.forEach { environment ->
-            var hasPin = false
-            var versionIsPending = false
-            val pendingVersionsToCheck: MutableSet<String> =
-              when (environment.constraints.anyStateful) {
-                true -> repository
-                  .pendingConstraintVersionsFor(deliveryConfig.name, environment.name)
-                  .filter { versions.contains(it) }
-                  .toMutableSet()
-                false -> mutableSetOf()
-              }
-
-            val queuedForApproval: MutableSet<String> =
-              when (environment.constraints.anyStateful) {
-                true -> repository
-                  .getQueuedConstraintApprovals(deliveryConfig.name, environment.name)
-                  .toMutableSet()
-                false -> mutableSetOf()
-              }
-
-            val vetoedVersions: Set<String> =
-              (vetoedArtifacts[envPinKey(environment.name, artifact)]?.versions)
-                ?: emptySet()
+          return@forEach
+        }
+        deliveryConfig.environments.forEach { environment ->
+          // if pinned, we approve that version right away without doing anything else
+          // to fast track that deployment
+          if (pinnedEnvs.hasPinFor(environment.name, artifact)) {
+            // todo: https://github.com/spinnaker/keel/issues/1254
+            val pinnedVersion: String = pinnedEnvs.versionFor(environment.name, artifact) ?: return@forEach
```

Comment:
Does this PR introduce a change in behavior with respect to pinning? This code looks different from what was here before, and doesn't look explicitly related to the refactor.

Reply:
Nope! It does not include any changes in behavior. It's just different because before it was very confusing.

Reply:
well, it shouldn't include any changes. 🤞  i'm not lying to you